File: src/sql_assistant/web/static/js/export.js
/**
 * Export Module - 导出查询结果
 * 支持 CSV、JSON、Excel 格式导出
 */

class ExportManager {
    /**
     * 导出为 CSV 格式
     * @param {Array} columns - 列名数组
     * @param {Array} rows - 数据行数组
     * @param {string} filename - 文件名（不含扩展名）
     */
    static exportToCSV(columns, rows, filename = 'query_result') {
        // 处理特殊字符，添加 BOM 支持中文
        const BOM = '\uFEFF';
        
        // 构建 CSV 内容
        let csv = BOM;
        
        // 添加表头
        csv += columns.map(col => this._escapeCSV(String(col))).join(',') + '\n';
        
        // 添加数据行
        for (const row of rows) {
            csv += row.map(cell => this._escapeCSV(String(cell ?? ''))).join(',') + '\n';
        }
        
        // 下载文件
        this._downloadFile(csv, `${filename}.csv`, 'text/csv;charset=utf-8');
    }

    /**
     * 导出为 JSON 格式
     * @param {Array} columns - 列名数组
     * @param {Array} rows - 数据行数组
     * @param {string} filename - 文件名（不含扩展名）
     */
    static exportToJSON(columns, rows, filename = 'query_result') {
        // 转换为对象数组格式
        const data = rows.map(row => {
            const obj = {};
            columns.forEach((col, index) => {
                obj[col] = row[index] ?? null;
            });
            return obj;
        });
        
        // 添加元数据
        const result = {
            export_time: new Date().toISOString(),
            columns: columns,
            row_count: rows.length,
            data: data
        };
        
        const jsonContent = JSON.stringify(result, null, 2);
        this._downloadFile(jsonContent, `${filename}.json`, 'application/json');
    }

    /**
     * 导出为 Excel 格式（使用 HTML table 方式）
     * @param {Array} columns - 列名数组
     * @param {Array} rows - 数据行数组
     * @param {string} filename - 文件名（不含扩展名）
     */
    static exportToExcel(columns, rows, filename = 'query_result') {
        // 创建 HTML table
        let html = `
<html xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>
    <meta charset="UTF-8">
    <style>
        table { border-collapse: collapse; }
        th, td { border: 1px solid #ccc; padding: 8px; }
        th { background: #f0f0f0; font-weight: bold; }
    </style>
</head>
<body>
    <table>
        <thead>
            <tr>${columns.map(col => `<th>${this._escapeHtml(String(col))}</th>`).join('')}</tr>
        </thead>
        <tbody>
            ${rows.map(row => `<tr>${row.map(cell => `<td>${this._escapeHtml(String(cell ?? ''))}</td>`).join('')}</tr>`).join('')}
        </tbody>
    </table>
</body>
</html>`;
        
        // 添加 Excel 特定的 MIME 类型
        const blob = new Blob([html], { 
            type: 'application/vnd.ms-excel;charset=utf-8' 
        });
        
        this._downloadBlob(blob, `${filename}.xls`);
    }

    /**
     * 显示导出格式选择菜单
     * @param {HTMLElement} anchorEl - 锚点元素（按钮位置）
     * @param {Array} columns - 列名数组
     * @param {Array} rows - 数据行数组
     */
    static showExportMenu(anchorEl, columns, rows) {
        // 移除已存在的菜单
        const existingMenu = document.querySelector('.export-menu');
        if (existingMenu) {
            existingMenu.remove();
        }
        
        // 创建菜单
        const menu = document.createElement('div');
        menu.className = 'export-menu';
        menu.innerHTML = `
            <div class="export-menu-header">选择导出格式</div>
            <button class="export-menu-item" data-format="csv">
                <span>📄</span> CSV
            </button>
            <button class="export-menu-item" data-format="json">
                <span>📋</span> JSON
            </button>
            <button class="export-menu-item" data-format="excel">
                <span>📊</span> Excel
            </button>
        `;
        
        // 获取锚点元素位置
        const rect = anchorEl.getBoundingClientRect();
        menu.style.position = 'fixed';
        menu.style.top = `${rect.top - menu.offsetHeight - 8}px`;
        menu.style.right = `${window.innerWidth - rect.right + 8}px`;
        
        // 添加点击事件
        menu.querySelectorAll('.export-menu-item').forEach(btn => {
            btn.addEventListener('click', (e) => {
                const format = e.currentTarget.dataset.format;
                const timestamp = new Date().toISOString().slice(0, 19).replace(/:/g, '-');
                const filename = `query_result_${timestamp}`;
                
                switch (format) {
                    case 'csv':
                        this.exportToCSV(columns, rows, filename);
                        break;
                    case 'json':
                        this.exportToJSON(columns, rows, filename);
                        break;
                    case 'excel':
                        this.exportToExcel(columns, rows, filename);
                        break;
                }
                
                menu.remove();
            });
        });
        
        // 点击外部关闭菜单
        document.addEventListener('click', function closeMenu(e) {
            if (!menu.contains(e.target) && e.target !== anchorEl) {
                menu.remove();
                document.removeEventListener('click', closeMenu);
            }
        });
        
        document.body.appendChild(menu);
        
        // 自动聚焦到菜单
        menu.focus();
    }

    /**
     * 转义 CSV 特殊字符
     * @param {string} value - 要转义的值
     * @returns {string} 转义后的值
     */
    static _escapeCSV(value) {
        // 如果包含逗号、引号或换行符，需要用引号包裹并转义内部引号
        if (value.includes(',') || value.includes('"') || value.includes('\n')) {
            return `"${value.replace(/"/g, '""')}"`;
        }
        return value;
    }

    /**
     * 转义 HTML 特殊字符
     * @param {string} value - 要转义的值
     * @returns {string} 转义后的值
     */
    static _escapeHtml(value) {
        const div = document.createElement('div');
        div.textContent = value;
        return div.innerHTML;
    }

    /**
     * 下载文件（文本内容）
     * @param {string} content - 文件内容
     * @param {string} filename - 文件名
     * @param {string} mimeType - MIME类型
     */
    static _downloadFile(content, filename, mimeType) {
        const blob = new Blob([content], { type: mimeType });
        this._downloadBlob(blob, filename);
    }

    /**
     * 下载 Blob 文件
     * @param {Blob} blob - Blob 对象
     * @param {string} filename - 文件名
     */
    static _downloadBlob(blob, filename) {
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        URL.revokeObjectURL(url);
    }
}

// 暴露到全局
window.ExportManager = ExportManager;

File: src/sql_assistant/api/export.py
"""导出相关 API 路由"""

import csv
import json
from datetime import datetime
from io import StringIO
from typing import List, Dict, Any

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from .models import ExportRequest, ExportResponse

router = APIRouter()


def _convert_to_csv(columns: List[str], rows: List[List[Any]]) -> str:
    """将数据转换为CSV格式"""
    output = StringIO()
    writer = csv.writer(output)
    
    # 写入表头
    writer.writerow(columns)
    
    # 写入数据行
    for row in rows:
        writer.writerow([str(cell) if cell is not None else '' for cell in row])
    
    return output.getvalue()


def _convert_to_json(columns: List[str], rows: List[List[Any]]) -> str:
    """将数据转换为JSON格式"""
    data = []
    for row in rows:
        obj = {}
        for i, col in enumerate(columns):
            obj[col] = row[i]
        data.append(obj)
    
    result = {
        "export_time": datetime.now().isoformat(),
        "columns": columns,
        "row_count": len(rows),
        "data": data
    }
    
    return json.dumps(result, ensure_ascii=False, indent=2)


def _convert_to_excel(columns: List[str], rows: List[List[Any]]) -> str:
    """将数据转换为Excel格式（HTML表格方式）"""
    html = f'''
<html xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>
    <meta charset="UTF-8">
    <style>
        table {{ border-collapse: collapse; width: 100%; }}
        th, td {{ border: 1px solid #ccc; padding: 8px; text-align: left; }}
        th {{ background: #f0f0f0; font-weight: bold; }}
    </style>
</head>
<body>
    <table>
        <thead>
            <tr>{''.join(f'<th>{col}</th>' for col in columns)}</tr>
        </thead>
        <tbody>
            {''.join(f'<tr>{''.join(f"<td>{str(cell) if cell is not None else ''}</td>" for cell in row)}</tr>' for row in rows)}
        </tbody>
    </table>
</body>
</html>'''
    return html


@router.post("/export", response_model=ExportResponse)
async def export_data(request: ExportRequest):
    """
    导出查询结果
    
    支持将查询结果导出为 CSV、JSON、Excel 格式。
    可用于前端无法处理的大数据量导出，或需要服务器端生成文件的场景。
    """
    try:
        if not request.columns or not request.rows:
            raise HTTPException(status_code=400, detail="导出数据不能为空")
        
        format_type = request.format.lower()
        
        if format_type == 'csv':
            content = _convert_to_csv(request.columns, request.rows)
            content_type = 'text/csv;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.csv'
        elif format_type == 'json':
            content = _convert_to_json(request.columns, request.rows)
            content_type = 'application/json;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.json'
        elif format_type == 'excel':
            content = _convert_to_excel(request.columns, request.rows)
            content_type = 'application/vnd.ms-excel;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.xls'
        else:
            raise HTTPException(status_code=400, detail=f"不支持的导出格式: {request.format}")
        
        return ExportResponse(
            success=True,
            format=format_type,
            content=content,
            content_type=content_type,
            filename=filename,
            row_count=len(request.rows)
        )
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"导出失败: {str(e)}")


@router.post("/export/download")
async def export_download(request: ExportRequest):
    """
    导出查询结果并直接下载
    
    支持流式下载，适合大数据量导出。
    """
    try:
        if not request.columns or not request.rows:
            raise HTTPException(status_code=400, detail="导出数据不能为空")
        
        format_type = request.format.lower()
        
        if format_type == 'csv':
            content = _convert_to_csv(request.columns, request.rows)
            content_type = 'text/csv;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.csv'
        elif format_type == 'json':
            content = _convert_to_json(request.columns, request.rows)
            content_type = 'application/json;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.json'
        elif format_type == 'excel':
            content = _convert_to_excel(request.columns, request.rows)
            content_type = 'application/vnd.ms-excel;charset=utf-8'
            filename = f'query_result_{datetime.now().strftime("%Y%m%d_%H%M%S")}.xls'
        else:
            raise HTTPException(status_code=400, detail=f"不支持的导出格式: {request.format}")
        
        # 添加 BOM 以支持 Excel 正确识别中文
        if format_type == 'csv':
            content = '\uFEFF' + content
        
        return StreamingResponse(
            iter([content]),
            media_type=content_type,
            headers={
                "Content-Disposition": f'attachment; filename="{filename}"',
                "Content-Type": content_type
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"导出失败: {str(e)}")